### PR TITLE
kwd: improvements

### PR DESCRIPTION
**Pull request by mrajwa**
This patch adds some improvements aimed to extend KPB logging
by providing informations like kpb state or timeout duration.

 Signed-off-by: Marcin Rajwa <marcin.rajwa@linux.intel.com>

### CODE REVIEW

File: src/audio/kpb.c [hunk 1]
```diff
@@ -639,6 +639,8 @@ static int kpb_buffer_data(struct comp_dev *dev, struct comp_buffer *source,
 	if (kpb->state == KPB_STATE_DRAINING)
 		draining_data->buffered_while_draining += size_to_copy;
 
+	timeout = platform_timer_get(platform_timer) +
```
Comment (tlauda) [inline src/audio/kpb.c]: Is it supposed to be 1 ms?

Comment (mrajwa) [reply src/audio/kpb.c]: Do you think it is too much? If buffering was interrupted that routine can take some time, so if we set timeout of 100us then I am afraid we will hit timeout frequently.

Comment (tlauda) [reply src/audio/kpb.c]: I was afraid it can be too short, but if the tests show otherwise, then it's OK.

**Comment by tlauda** [comment]
@mrajwa Can you fix the typo in second commit message, co we could merge this?

**Comment by mrajwa** [comment]
@tlauda @lgirdwood @jajanusz , PR updated with new patches.

File: src/audio/kpb.c [hunk 2]
```diff
@@ -1252,6 +1256,23 @@ static inline bool validate_host_params(size_t host_period_size,
 
 	return (drained_per_interval > host_period_size) ? true : false;
 }
+
+/**
+ * \brief Change KPB state and log this change internaly.
```
Comment (tlauda) [inline src/audio/kpb.c]: Typo: internaly.

File: src/audio/kpb.c [hunk 3]
```diff
@@ -43,6 +43,7 @@
 
 /* KPB private data, runtime data */
 struct comp_data {
+	uint64_t state_log;
```
Comment (tlauda) [inline src/audio/kpb.c]: Do we need 64 bits?

File: src/audio/kpb.c [hunk 4]
```diff
@@ -344,7 +344,8 @@ static int kpb_prepare(struct comp_dev *dev)
 	if (kpb->state == KPB_STATE_RESETTING ||
 	    kpb->state == KPB_STATE_RESET_FINISHING) {
 		trace_kpb_error("kpb_prepare() error: can not prepare KPB "
-				"due to ongoing reset.");
+				"due to ongoing reset. state log %d",
```
Comment (tlauda) [inline src/audio/kpb.c]: Commit description typo: serios.

File: src/audio/kpb.c [hunk 5]
```diff
@@ -602,8 +604,9 @@ static int kpb_copy(struct comp_dev *dev)
 		ret = PPL_STATUS_PATH_STOP;
 		break;
 	default:
-		trace_kpb_error("kpb_copy(): wrong state - %d! "
-				" Copy forbidden.", kpb->state);
+		trace_kpb_error("kpb_copy(): wrong state, copy forbidden. "
+				"(state %d, state log %d) ",
```
Comment (tlauda) [inline src/audio/kpb.c]: Space at the end not needed.

File: src/audio/kpb.c (hunk 4 above)
Comment (tlauda) [reply src/audio/kpb.c]: Comma instead of period?

File: src/audio/kpb.c [hunk 6]
```diff
@@ -424,6 +426,13 @@ static int kpb_prepare(struct comp_dev *dev)
 		}
 	}
 
+	if (!kpb->sel_sink || !kpb->host_sink) {
+		trace_kpb("kpb_prepare() error: could not find "
+			  "sinks: sel_sink %d host_sink %d",
+			  (uint32_t)kpb->sel_sink, (uint32_t)kpb->host_sink);
```
Comment (tlauda) [inline src/audio/kpb.c]: This is potential exception, when one of them will be NULL.

File: src/audio/kpb.c [hunk 7]
```diff
@@ -466,39 +466,48 @@ static void kpb_cache(struct comp_dev *dev, int cmd)
...
+	switch (kpb->state) {
+	case KPB_STATE_BUFFERING:
+	case KPB_STATE_DRAINING:
 		/* KPB is performing some task now,
 		 * terminate it gently.
 		 */
 		kpb_change_state(kpb, KPB_STATE_RESETTING);
-		return -EBUSY;
-	}
+		ret = -EBUSY;
```
Comment (tlauda) [inline src/audio/kpb.c]: break is missing.

File: src/audio/kpb.c [hunk 8]
```diff
@@ -466,39 +466,48 @@ static void kpb_cache(struct comp_dev *dev, int cmd)
...
+		/* Unregister KPB for async notification */
+		notifier_unregister(&kpb->kpb_events);
+
+		/* Finally KPB is ready after reset */
+		kpb_change_state(kpb, KPB_STATE_PREPARING);
 
-	return comp_set_state(dev, COMP_TRIGGER_RESET);
+		ret = comp_set_state(dev, COMP_TRIGGER_RESET);
+		break;
+
```
Comment (tlauda) [inline src/audio/kpb.c]: Empty line not needed.

File: src/audio/kpb.c [hunk 9]
```diff
@@ -557,6 +557,14 @@ static int kpb_copy(struct comp_dev *dev)
 		}
 
 		copy_bytes = MIN(sink->free, source->avail);
+		if (!copy_bytes) {
+			trace_kpb_error("kpb_copy() error: nothing to copy "
+					"sink->free %d sorce->avail %d",
```
Comment (tlauda) [inline src/audio/kpb.c]: Typo: sorce.

File: src/audio/kpb.c [hunk 10]
```diff
@@ -557,6 +557,14 @@ static int kpb_copy(struct comp_dev *dev)
 		}
 
 		copy_bytes = MIN(sink->free, source->avail);
+		if (!copy_bytes) {
```
Comment (tlauda) [inline src/audio/kpb.c]: The last sentence of the commit description seems weird.

File: src/audio/kpb.c [hunk 11]
```diff
@@ -595,12 +603,20 @@ static int kpb_copy(struct comp_dev *dev)
...
+			trace_kpb_error("kpb_copy() error: nothing to copy "
+					"sink->free %d sorce->avail %d",
```
Comment (tlauda) [inline src/audio/kpb.c]: Typo: sorce.

File: src/audio/kpb.c [hunk 12]
```diff
@@ -595,12 +603,20 @@ static int kpb_copy(struct comp_dev *dev)
...
+			goto out;
+		}
+
 		kpb_copy_samples(sink, source, copy_bytes, sample_width);
 
 		comp_update_buffer_produce(sink, copy_bytes);
 		comp_update_buffer_consume(source, copy_bytes);
 
-		ret = 0;
+		ret = PPL_STATUS_PATH_STOP;
```
Comment (tlauda) [inline src/audio/kpb.c]: Is it intentional change?

File: src/audio/kpb.c (hunk 6 above)
Comment (jajanusz) [reply src/audio/kpb.c]: @tlauda what do you mean? If any sink is null it will just be casted to 0

Comment (tlauda) [reply src/audio/kpb.c]: Of course. I don't know what I was thinking...

File: src/audio/kpb.c (hunk 3 above)
Comment (mrajwa) [reply src/audio/kpb.c]: Wider data size gives us the possibility to log more states. So I would keep it as 64 bits.

File: src/audio/kpb.c (hunk 12 above)
Comment (mrajwa) [reply src/audio/kpb.c]: Yes, during host copy we no longer need selector and detector components.

File: src/audio/kpb.c (hunk 10 above)
Comment (mrajwa) [reply src/audio/kpb.c]: Yes, corrected now.

**Comment by mrajwa** [comment]
@tlauda, updated.

**Comment by tlauda** [comment]
@mrajwa Can you fix the checkpatch?

**Comment by mrajwa** [comment]
@tlauda, sure I have just did.

**Pull request merged**